meta-refkit/classes/image-installer.bbclass

```diff
@@ -0,0 +1,370 @@
+        try:
+            os.link(path, dest)
+        except IOError as ex:
+            bb.warn('Hard-linking from %s to %s failed. Falling back to full copy, which looses sparseness: %s' %
+                    (path, dest, str(ex)))
+            shutil.copy2(path, dest)
+        # We run under pseudo, so chown/chmod does not really change the attributes
+        # of the original image file.
+        os.chown(dest, 0, 0)
+        mode = stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH
+        os.chmod(dest, mode)
+}
+do_rootfs[depends] += "${@ ' '.join([x.split(':')[0] + ':do_image_complete' for x in '${INSTALLER_SOURCE_IMAGES}'.split()]) }"
+do_rootfs[postfuncs] += " install_images "
+
+
+# The default installer can be used interactively or
+# configured to run automatically by setting environment
+# variables in advance.
+#
+# The flow is:
+# - determine which image is to be installed
+# - determine which disks are suitable as target, automatically
+#   filtering out the boot disk and everything that is read-only
+#   or completely unusable (removable media)
+# - determine which target to install to
+# - execute installation
+#
+# See below for the commands used by each step.
+#
+# The default installer code is free of bashisms and passed
+# checks with checkbashisms.pl and checkshell. The "local"
+# keyword (although supported by bash and dash) is avoided
+# in favor of wrapping function bodies with local variables
+# in a subshell.
+INSTALLER_DEFAULT () {
+#!/bin/sh -e
+
+${INSTALLER_LOGGING}
+${INSTALLER_UTILS}
+${INSTALLER_PICK_INPUT}
+${INSTALLER_FIND_OUTPUT}
+${INSTALLER_PICK_OUTPUT}
+${INSTALLER_INSTALL}
+
+pick_input
+find_output
+pick_output
+install_image
+info "Installed $CHOSEN_INPUT on $CHOSEN_OUTPUT successfully."
+}
+
+INSTALLER ??= "${INSTALLER_DEFAULT}"
+
+# Ensure that whatever runtime tools are needed by the installer
+# script are actually in the image.
+INSTALLER_RDEPENDS ??= " \
+    util-linux \
+"
+IMAGE_INSTALL_append = " ${INSTALLER_RDEPENDS}"
+
+# Installs whatever is contained in ${INSTALLER} as ${INSTALLER_BINARY}.
+python install_installer () {
+    import stat
+
+    destfile = d.getVar('INSTALLER_BINARY')
+    bb.utils.mkdirhier(os.path.dirname(destfile))
+    with open(destfile, 'w') as f:
+        f.write(d.getVar('INSTALLER'))
+    os.chown(destfile, 0, 0)
+    mode = stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH | \
+        stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH
+    os.chmod(destfile, mode)
+}
+do_rootfs[postfuncs] += " install_installer "
+
+# Utility code for reporting to the user of the script.
+INSTALLER_LOGGING[shellcheck] = "sh"
+INSTALLER_LOGGING () {
+    fatal () {
+        echo >&2 "ERROR: $*"
+        exit 1
+    }
+
+    error () {
+        echo >&2 "ERROR: $*"
+    }
+
+    info () {
+        echo >/dev/tty "$*"
+    }
+}
+
+# Cleanup handling, common code for selecting among different input and output
+# options interactively, etc.
+INSTALLER_UTILS[shellcheck] = "sh"
```

Comment:
Did you run this through shellcheck? It reports ~8 warnings and style issues, nothing major though.